In [1]:
# imports 

import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [2]:
data = pd.read_excel('../data/Cat_personality_data.xlsx')

In [3]:
data

Personality1_Vigilant  Personality2_Stable  Personality3_Bold  \
0                         6                    5                  7   
1                         7                    3                  4   
2                         7                    5                  7   
3                         7                    4                  1   
4                         5                    7                  7   
...                     ...                  ...                ...   
2797                      5                    6                  4   
2798                      5                    7                  6   
2799                      6                    5                  5   
2800                      6                    6                  5   
2801                      7                    6                  5   

      Personality4_Clumsy  Personality5_Defiant  Personality6_Gentle  \
0                       5                     5                    6   
1                       1                     7                    7   
2                       1                     2                    3   
3                       1                     7                    2   
4                       1                     4                    7   
...                   ...                   ...                  ...   
2797                    1                     1                    3   
2798                    1                     2                    6   
2799                    1                     2                    6   
2800                    1                     1                    7   
2801                    2                     2                    6   

      Personality7_Constrained  Personality8_Inquisitive  \
0                            1                         5   
1                            3                         4   
2                            3                         7   
3                            7                         1   
4                            7                         4   
...                        ...                       ...   
2797                         5                         6   
2798                         6                         3   
2799                         2                         5   
2800                         4                         5   
2801                         6                         6   

      Personality9_Inventive  Personality10_Irritable  ...  \
0                          4                        3  ...   
1                          3                        2  ...   
2                          7                        2  ...   
3                          3                        7  ...   
4                          4                        1  ...   
...                      ...                      ...  ...   
2797                       4                        2  ...   
2798                       2                        1  ...   
2799                       5                        1  ...   
2800                       6                        2  ...   
2801                       4                        2  ...   

      Personality45_Decisive  Personality46_Self_assured  \
0                          4                           6   
1                          4                           3   
2                          7                           6   
3                          5                           3   
4                          5                           7   
...                      ...                         ...   
2797                       6                           6   
2798                       5                           6   
2799                       5                           5   
2800                       6                           6   
2801                       5                           6   

      Personality47_Anxious  Personality48_Trusting  Personality49_Active  \
0                       

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2802 entries, 0 to 2801
Data columns (total 54 columns):
 #   Column                               Non-Null Count  Dtype 
---  ------                               --------------  ----- 
 0   Personality1_Vigilant                2802 non-null   int64 
 1   Personality2_Stable                  2802 non-null   int64 
 2   Personality3_Bold                    2802 non-null   int64 
 3   Personality4_Clumsy                  2802 non-null   int64 
 4   Personality5_Defiant                 2802 non-null   int64 
 5   Personality6_Gentle                  2802 non-null   int64 
 6   Personality7_Constrained             2802 non-null   int64 
 7   Personality8_Inquisitive             2802 non-null   int64 
 8   Personality9_Inventive               2802 non-null   int64 
 9   Personality10_Irritable              2802 non-null   int64 
 10  Personality11_Distractable           2802 non-null   int64 
 11  Personality12_Erratic                2802 n

In [5]:
data.describe()

Personality1_Vigilant  Personality2_Stable  Personality3_Bold  \
count            2802.000000          2802.000000        2802.000000   
mean                5.491792             4.744111           4.387580   
std                 1.420801             1.722183           1.806914   
min                 1.000000             1.000000           1.000000   
25%                 5.000000             3.000000           3.000000   
50%                 6.000000             5.000000           5.000000   
75%                 7.000000             6.000000           6.000000   
max                 7.000000             7.000000           7.000000   

       Personality4_Clumsy  Personality5_Defiant  Personality6_Gentle  \
count          2802.000000           2802.000000          2802.000000   
mean              2.448965              3.738044             5.276588   
std               1.766360              1.995508             1.632262   
min               1.000000              1.000000             1.000000   
25%               1.000000              2.000000             4.000000   
50%               2.000000              4.000000             6.000000   
75%               4.000000              5.000000             7.000000   
max               7.000000              7.000000             7.000000   

       Personality7_Constrained  Personality8_Inquisitive  \
count               2802.000000               2802.000000   
mean                   4.012491                  5.214847   
std                    1.628092                  1.603355   
min                    1.000000                  1.000000   
25%                    3.000000                  4.000000   
50%                    4.000000                  5.000000   
75%                    5.000000                  7.000000   
max                    7.000000                  7.000000   

       Personality9_Inventive  Personality10_Irritable  ...  \
count             2802.000000              2802.000000  ...   
mean                 4.801570                 2.675232  ...   
std                  1.794481                 1.807049  ...   
min                  1.000000                 1.000000  ...   
25%                  3.250000                 1.000000  ...   
50%                  5.000000                 2.000000  ...   
75%                  6.000000                 4.000000  ...   
max                  7.000000                 7.000000  ...   

       Personality43_Playful  Personality44_Vocal  Personality45_Decisive  \
count            2802.000000          2802.000000             2802.000000   
mean                5.185582             5.251963                5.158815   
std                 1.527038             1.784150                1.325919   
min                 1.000000             1.000000                1.000000   
25%                 5.000000             4.000000                4.000000   
50%                 5.000000             6.000000                5.000000   
75%                 6.000000             7.000000                6.000000   
max                 7.000000             7.000000                7.000000   

       Personality46_Self_assured  Personality47_Anxious  \
count                 2802.000000            2802.000000   
mean                     5.362598               3.086367   
std                      1.433470               1.743770   
min                      1.000000               1.000000   
25%                      5.000000               2.000000   
50%                      6.000000               3.000000   
75%                      6.000000               4.000000   
max                      7.000000               7.000000   

       Personality48_Trusting  Personality49_Active  \
count             2802.000000           2802.000000   
mean                 4.255532              5.160956   
std                  1.875169              1.705820   
min                  1.000000              1.000000   
25%                  3.000000              4.000000   
50%              

### Renaming personality trait columns for easier readibility 

In [6]:
data.rename(columns=lambda col: col.split('_', 1)[1] if '_' in col else col, inplace=True)
data.rename(columns={'sex': 'Cat_sex'}, inplace=True)

### Removing null values

In [8]:
data[data['Cat_sex'].isnull() == True]

Vigilant  Stable  Bold  Clumsy  Defiant  Gentle  Constrained  \
45           6       6     6       2        6       5            6   
86           5       5     4       1        5       4            5   
476          4       3     5       7        2       7            4   
500          4       6     6       1        4       3            4   
670          5       7     5       5        3       6            5   
789          6       5     2       3        3       6            4   
961          7       7     4       1        4       7            6   
969          5       4     7       1        6       6            5   
1329         4       5     6       1        6       2            4   
1345         5       7     6       1        6       5            5   
1457         1       4     4       5        3       7            6   
1488         7       3     3       5        6       5            3   
1502         7       5     7       1        5       7            7   
1520         6       3     2       3        3       6            6   
1869         6       3     5       1        4       5            3   
1876         7       7     7       1        7       4            5   
2029         4       5     3       5        6       1            3   
2072         5       5     5       3        3       6            5   
2131         7       5     6       1        4       1            4   
2267         6       5     4       4        1       7            5   
2445         7       7     5       6        6       5            6   
2479         4       3     3       1        2       4            5   
2511         4       3     2       3        2       6            5   

      Inquisitive  Inventive  Irritable  ...  Decisive  Self_assured  Anxious  \
45              5          6          3  ...         5             5        3   
86              6          6          6  ...         4             6        2   
476             5          3          1  ...         5             5        3   
500             3          4          4  ...         5             5        3   
670             7          6          1  ...         5             6        2   
789             6          6          3  ...         6             5        3   
961             7          7          1  ...         5             7        1   
969             7          7          5  ...         7             7        1   
1329            6          5          5  ...         5             5        1   
1345            5          5          1  ...         6             7        2   
1457            2          1          1  ...         2             7        2   
1488            5          3          6  ...         6             5        4   
1502            7          7          2  ...         7             7        1   
1520            2          2          4  ...         4             2        6   
1869            7          7          2  ...         6             6        4   
1876            5          4          1  ...         7             7        1   
2029            2          1          5  ...         6             6        4   
2072            5          5          3  ...         3             5        3   
2131            4          4          1  ...         7             7        1   
2267            5          3          2  ...         4             6        2   
2445            5          5          2  ...         5             5        5   
2479            3          3          5  ...         5             7        6   
2511            4          3          2  ...         4             3        6   

      Trusting  Active  Cooperative  Shy  Eccentric     Country  Cat_sex  
45           5       3            4    4          2   Australia      NaN  
86           4       7            2    2          4  NewZealand      NaN  
476          6       3            6    3          5   Australia      NaN  
500          2       4            3    2          4   Australia      NaN  
670  

Only the `Cat_sex` column has null values, and there are only 23 instances so they will be removed. 

In [9]:
data = data.dropna()

In [10]:
data

Vigilant  Stable  Bold  Clumsy  Defiant  Gentle  Constrained  \
0            6       5     7       5        5       6            1   
1            7       3     4       1        7       7            3   
2            7       5     7       1        2       3            3   
3            7       4     1       1        7       2            7   
4            5       7     7       1        4       7            7   
...        ...     ...   ...     ...      ...     ...          ...   
2797         5       6     4       1        1       3            5   
2798         5       7     6       1        2       6            6   
2799         6       5     5       1        2       6            2   
2800         6       6     5       1        1       7            4   
2801         7       6     5       2        2       6            6   

      Inquisitive  Inventive  Irritable  ...  Decisive  Self_assured  Anxious  \
0               5          4          3  ...         4             6        4   
1               4          3          2  ...         4             3        6   
2               7          7          2  ...         7             6        3   
3               1          3          7  ...         5             3        3   
4               4          4          1  ...         5             7        1   
...           ...        ...        ...  ...       ...           ...      ...   
2797            6          4          2  ...         6             6        2   
2798            3          2          1  ...         5             6        1   
2799            5          5          1  ...         5             5        2   
2800            5          6          2  ...         6             6        1   
2801            6          4          2  ...         5             6        2   

      Trusting  Active  Cooperative  Shy  Eccentric     Country  Cat_sex  
0            5       7            2    2          4  NewZealand     Male  
1            6       1            6    2          5  NewZealand   Female  
2            4       7            2    3          4   Australia     Male  
3            2       2            2    6          2   Australia   Female  
4            7       7            6    2          7   Australia     Male  
...        ...     ...          ...  ...        ...         ...      ...  
2797         4       7            5    3          1  NewZealand     Male  
2798         6       1            3    4          1  NewZealand   Female  
2799         5       5            4    2          4  NewZealand   Female  
2800         2       5            5    5          3  NewZealand   Female  
2801         6       6            5    4          2  NewZealand     Male  

[2779 rows x 54 columns]

### Checking values for each column
#### Expected values:
 - The 52 personality traits should be integers ranging from 1-7
 - Country should be New Zealand or Australia 
 - Cat_sex should be male or female 

In [11]:
for column in data.columns:
    print(data[column].unique())

[6 7 5 4 1 2 3]
[5 3 4 7 2 1 6]
[7 4 1 5 6 3 2]
[5 1 6 4 2 3 7]
[5 7 2 4 6 1 3]
[6 7 3 2 5 4 1]
[1 3 7 6 2 5 4]
[5 4 7 1 3 6 2]
[4 3 7 1 5 6 2]
[3 2 7 1 4 5 6]
[3 5 2 1 6 4 7]
[1 2 4 6 5 3 7]
[2 4 7 6 5 1 3]
[5 3 1 2 4 6 7]
[4 2 5 3 6 1 7]
[5 3 7 4 1 6 2]
[5 3 7 6 2 4 1]
[5 2 7 1 3 4 6]
[6 1 3 2 5 7 4]
[5 3 7 4 6 1 2]
[5 7 2 4 3 1 6]
[5 2 4 3 1 7 6]
[6 1 3 2 4 5 7]
[5 6 7 4 2 1 3]
[5 6 2 3 1 7 4]
[3 6 2 7 4 5 1]
[6 1 3 7 2 5 4]
[4 2 5 1 7 6 3]
[5 7 2 3 6 1 4]
[2 6 4 5 1 3 7]
[4 6 2 1 5 3 7]
[5 4 7 2 3 1 6]
[2 4 6 3 5 7 1]
[5 7 6 2 1 3 4]
[4 6 2 1 5 7 3]
[3 1 6 4 2 7 5]
[3 2 5 7 6 4 1]
[1 2 4 5 6 7 3]
[5 2 3 7 6 4 1]
[5 7 3 2 4 1 6]
[5 4 2 6 1 3 7]
[5 7 3 1 2 6 4]
[4 7 2 5 3 1 6]
[7 2 6 5 1 3 4]
[4 7 5 3 1 6 2]
[6 3 7 5 2 1 4]
[4 6 3 1 2 7 5]
[5 6 4 2 7 1 3]
[7 1 2 3 6 5 4]
[2 6 4 5 3 7 1]
[2 3 6 4 5 7 1]
[4 5 2 7 1 3 6]
['NewZealand' 'Australia']
['Male' 'Female' 'Unsure']


In [12]:
# Inspecting the datapoints where the cat sex is 'unsure'

data[data['Cat_sex'] == 'Unsure']

Vigilant  Stable  Bold  Clumsy  Defiant  Gentle  Constrained  \
35           2       3     1       4        2       6            6   
373          7       6     6       1        7       7            4   
581          7       6     6       4        6       5            5   
595          2       3     1       2        4       3            3   
772          3       6     6       1        5       6            2   
1364         6       3     2       2        5       5            4   
1665         7       7     6       1        4       6            6   
1837         7       4     6       1        7       4            6   
1847         7       2     1       4        1       4            6   
1973         6       3     2       1        2       6            3   
2273         6       4     4       3        7       5            4   
2344         7       7     7       2        4       7            2   
2485         7       7     7       1        2       4            4   
2785         7       6     7       2        5       7            1   
2796         7       7     5       1        6       7            4   

      Inquisitive  Inventive  Irritable  ...  Decisive  Self_assured  Anxious  \
35              3          3          6  ...         3             3        5   
373             5          6          4  ...         7             7        3   
581             6          5          4  ...         4             6        2   
595             2          4          2  ...         5             5        2   
772             7          7          1  ...         6             6        2   
1364            4          1          1  ...         6             4        6   
1665            3          3          1  ...         7             7        1   
1837            7          6          5  ...         6             4        4   
1847            3          2          2  ...         5             2        6   
1973            7          5          2  ...         5             6        3   
2273            5          6          7  ...         5             5        6   
2344            6          5          3  ...         5             6        3   
2485            3          1          1  ...         3             3        3   
2785            7          7          1  ...         6             7        1   
2796            7          7          1  ...         1             1        2   

      Trusting  Active  Cooperative  Shy  Eccentric     Country  Cat_sex  
35           5       2            5    5          4   Australia   Unsure  
373          3       5            5    5          4   Australia   Unsure  
581          6       3            2    1          6  NewZealand   Unsure  
595          2       2            3    3          1   Australia   Unsure  
772          7       7            5    1          1   Australia   Unsure  
1364         3       5            3    5          1  NewZealand   Unsure  
1665         7       7            7    1          1   Australia   Unsure  
1837         7       4            6    4          4   Australia   Unsure  
1847         1       5            1    7          3   Australia   Unsure  
1973         1       5            5    5          6  NewZealand   Unsure  
2273         2       4            3    5          4   Australia   Unsure  
2344         6       6            4    2          2   Australia   Unsure  
2485         1       1            4    4          2   Australia   Unsure  
2785         7       7            5    1          7   Australia   Unsure  
2796         6       1            7    3          1  NewZealand   Unsure  

[15 rows x 54 columns]

There are only 15 instances of the cat's sex being unsure, so those datapoints will be removed as well. 

In [13]:
data = data[data['Cat_sex'] != 'Unsure']

In [14]:
# Validating cleaned dataset

data

Vigilant  Stable  Bold  Clumsy  Defiant  Gentle  Constrained  \
0            6       5     7       5        5       6            1   
1            7       3     4       1        7       7            3   
2            7       5     7       1        2       3            3   
3            7       4     1       1        7       2            7   
4            5       7     7       1        4       7            7   
...        ...     ...   ...     ...      ...     ...          ...   
2797         5       6     4       1        1       3            5   
2798         5       7     6       1        2       6            6   
2799         6       5     5       1        2       6            2   
2800         6       6     5       1        1       7            4   
2801         7       6     5       2        2       6            6   

      Inquisitive  Inventive  Irritable  ...  Decisive  Self_assured  Anxious  \
0               5          4          3  ...         4             6        4   
1               4          3          2  ...         4             3        6   
2               7          7          2  ...         7             6        3   
3               1          3          7  ...         5             3        3   
4               4          4          1  ...         5             7        1   
...           ...        ...        ...  ...       ...           ...      ...   
2797            6          4          2  ...         6             6        2   
2798            3          2          1  ...         5             6        1   
2799            5          5          1  ...         5             5        2   
2800            5          6          2  ...         6             6        1   
2801            6          4          2  ...         5             6        2   

      Trusting  Active  Cooperative  Shy  Eccentric     Country  Cat_sex  
0            5       7            2    2          4  NewZealand     Male  
1            6       1            6    2          5  NewZealand   Female  
2            4       7            2    3          4   Australia     Male  
3            2       2            2    6          2   Australia   Female  
4            7       7            6    2          7   Australia     Male  
...        ...     ...          ...  ...        ...         ...      ...  
2797         4       7            5    3          1  NewZealand     Male  
2798         6       1            3    4          1  NewZealand   Female  
2799         5       5            4    2          4  NewZealand   Female  
2800         2       5            5    5          3  NewZealand   Female  
2801         6       6            5    4          2  NewZealand     Male  

[2764 rows x 54 columns]

The final cleaned dataset has 2764 rows and 54 columns, 38 rows were removed during preprocessing. The personality trait columns were also renamed to remove the `Personality#_` prefix for easier readibility

In [15]:
# Saving cleaned dataset to directory

current_dir = os.getcwd()
main_dir = os.path.abspath(os.path.join(current_dir, '..'))
data_dir = os.path.abspath(os.path.join(main_dir, 'data'))
data.to_csv(os.path.join(data_dir, 'data_preprocessed.csv'), index=False)